File: compiler/test/polynomial_domains.py
"""Admitted-source domain facts: exact expressions, hostile reception, coverage.

What the independent Lean consumer makes of the same source is a separate test,
tests/protocol/test_polynomial_domains_reference.py, because it needs a second build.
"""
import json
from commands import Commands
from tools import corpus, records



commands = Commands(records())


def run(command, text, *args, refuses=None):
    """What the compiler answered, read as JSON, or the refusal itself."""
    printed = commands.source(command, text, *args, refuses=refuses)
    return commands.last if refuses else json.loads(printed)


def inspect(text, pair=None):
    args = ['main'] + ([f'--compare={pair[0]},{pair[1]}'] if pair else [])
    return run('domain-inspect', text, *args)


def relation(source, left, right, expected, reason=None):
    c = inspect(source, (left, right))['comparison']
    assert c['relation'] == expected, c
    assert c['reasons'], c
    if reason:
        assert reason in c['reasons'], c


source = corpus.joinpath('polynomial-domains.pir').read_text().rstrip('\n')
report = inspect(source)
assert inspect(json.dumps(run('protocol-source', source))) == report
assert report['format'] == 'zkc.polynomial-domains/1'
domains = report['domains']
assert len(domains) == 10, domains
assert [d['association'] for d in domains].count('fold-result') == 1
assert all(d['field'] == 'koala-bear' for d in domains)
convention = domains[0]['convention']
assert convention['max_log_size'] == 24
assert convention['maximal_root'] == '1791270792'
assert convention['root_field'] == 'koala-bear'
assert all(d['convention'] == convention for d in domains)
for i in range(1, 6):
    relation(source, 0, i, 'same')
    assert domains[0]['congruence_class'] == domains[i]['congruence_class']
relation(source, 0, 6, 'different', 'size-positive-domain-versus-half')
relation(source, 6, 7, 'same')
relation(source, 6, 8, 'same')
relation(source, 0, 9, 'unknown', 'shift-equality-unproved')
assert [u['compatibility']['relation'] for u in report['uses']] == [
    'same', 'same', 'same', 'unknown']
terms = report['terms']
assert terms[domains[6]['size']]['kind'] == 'half'
assert terms[domains[6]['shift']]['kind'] == 'square'
for name in ('back', 'coefficients', 'scalar'):
    value = next(v for v in report['values'] if v['name'] == name)
    assert value['domain'] is None
    assert value['unknown_reason'] == 'no-installed-domain-transfer'
    assert value['may_depend_on']  # a dependency is not a coset identity

# Exact numeric constants from distinct operations; different shifts mean
# different ORDERED domains, without claiming the underlying sets are disjoint.
constants = '''module {
  fn Points<F: domain Field>()->() requires(TwoAdicField(F)) {
    let four = index::constant() attributes ("4"); let eight = index::constant() attributes ("8");
    let two = field::constant::<F>() attributes ("2"); let three = field::constant::<F>() attributes ("3");
    let again = field::constant::<F>() attributes ("2");
    let a = poly::domain_points::<F>(two,four);
    let b = poly::domain_points::<F>(three,four);
    let c = poly::domain_points::<F>(two,eight);
    let d = poly::domain_points::<F>(again,four);
    return ();
  }
  configure Base=Points(F=koala-bear);
  configure Extension=Points(F=koala-bear.ext8-binomial3);
  protocol Main { roles(P); local P:Base(); local P:Extension(); return(); }
  instance main_instance:Main {roles(P=P);} entry main=main_instance;
}'''
relation(constants, 0, 1, 'different', 'shift-distinct-canonical-constants')
relation(constants, 0, 2, 'different', 'size-distinct-canonical-constants')
relation(constants, 0, 3, 'same')
relation(constants, 0, 4, 'different', 'distinct-nominal-fields')
extended = inspect(constants)['domains'][4]
assert extended['field'] == 'koala-bear.ext8-binomial3'
assert extended['convention'] == convention

receptions = '''module {
  bind points=poly::domain_points(koala-bear);
  bind interpolate=poly::coset_interpolate(koala-bear);
  fn Points(s:koala-bear::Element,n:index)->(Vector<koala-bear::Element>) {
    let v = points(s,n); return(v);
  }
  fn Alias(s:koala-bear::Element)->(koala-bear::Element) {return(s);}
  fn Back(v:Vector<koala-bear::Element>,s:koala-bear::Element)->() {
    let p = interpolate(v,s); return();
  }
  protocol Main {
    roles(P,V); inputs(P s:koala-bear::Element,P n:index);
    local P:let sent = Points(s,n);
    message size:P(n)->V(n1);
    message first:P(s)->V(s1);
    message second:P(s)->V(s2);
    message word:P(sent)->V(received);
    local V:let a = Points(s1,n1);
    local V:let alias = Alias(s1);
    local V:let b = Points(alias,n1);
    local V:let c = Points(s2,n1);
    local V:Back(received,s1);
    return();
  }
  instance main_instance:Main {roles(P=P,V=V);} entry main=main_instance;
}'''
relation(receptions, 0, 1, 'unknown', 'size-equality-unproved')
relation(receptions, 1, 2, 'same')
relation(receptions, 1, 3, 'unknown', 'shift-equality-unproved')
r = inspect(receptions)
assert r['uses'][0]['compatibility'] == {
    'relation': 'unknown', 'reasons': ['reception-has-no-trusted-domain']}
received = [v for v in r['values'] if 'authored_sender_value' in v]
assert len(received) == 4
assert all(v['domain'] is None and not v['may_depend_on'] for v in received)
# Repeated authored output names in separate invocations never equate values.
assert len([v for v in r['values'] if v['name'] == 'v']) == 4

# Separate receptions of size also remain unknown with the exact same shift.
separate_size = receptions.replace('message second:P(s)->V(s2);',
    'message second:P(s)->V(s2); message size2:P(n)->V(n2);')
separate_size = separate_size.replace('Points(s2,n1)', 'Points(s1,n2)')
relation(separate_size, 1, 3, 'unknown', 'size-equality-unproved')

# Multilinear data gets no multiplicative coset by field, shape or name.
multilinear = '''module {
  bind fold=poly::fold(bls12-381.fr);
  fn Work(t:Table<"bls12-381.fr">,s:"bls12-381.fr"::Element)->(Table<"bls12-381.fr">) {
    let next = fold(t,s); return(next);
  }
  protocol Main {roles(P); inputs(P t:Table<"bls12-381.fr">,P s:"bls12-381.fr"::Element);
    local P:let v = Work(t,s); return();}
  instance main_instance:Main {roles(P=P);} entry main=main_instance;
}'''
m = inspect(multilinear)
assert not m['domains'] and not m['uses']
assert next(v for v in m['values'] if v['name'] == 'next')['unknown_reason'] == \
    'no-installed-domain-transfer'

# An arbitrary word can be interpreted repeatedly on the same coset without
# inheriting a low-degree or producer fact. The different interpretation is legal.
arbitrary = source.replace('let v = poly::coset_evaluate::<F>(p,s,n);',
    'let v = poly::coefficients::<F>(p);')
a = inspect(arbitrary)
assert a['uses'][0]['compatibility']['reasons'] == ['no-producer-domain-fact']
assert a['domains'][0]['size_origin_is_length']
assert a['terms'][a['domains'][0]['size']]['kind'] == 'length'

# Finite loop occurrences retain distinct reception leaves even when authored
# names and sender provenance are identical on each iteration.
loop = receptions.replace('    local V:let a = Points(s1,n1);', '''
    loop [repeated] 2 carry () capture(s,n1) -> () {
      message again:P(s)->V(fresh);
      local V:let word = Points(fresh,n1);
      yield();
    }
    local V:let a = Points(s1,n1);''')
relation(loop, 1, 2, 'unknown', 'shift-equality-unproved')

# Admission is mandatory. A missing algebraic requirement cannot be converted
# into an optimistic fact, even if a closed selection happens to be odd.
missing = source.replace(',CharacteristicNotTwo(F)', '')
run('domain-inspect', missing, 'main', refuses='generic-public-requirement')
run('domain-inspect', source.replace('koala-bear', 'bls12-381.fr'),
    'main', refuses='source-protocol-requirement')
run('domain-inspect', source, 'absent', refuses='execution-entry')
run('domain-inspect', source, 'main', '--compare=x,1', refuses='unsupported-option')
relation(source, 0, 1000, 'unknown', 'domain-id-out-of-range')

print(f'{commands.save()} polynomial domain inspection checks passed')

File: compiler/lib/Analysis/PolynomialDomains.cpp
#include "zkc/Analysis/PolynomialDomains.h"
#include "zkc/Contracts/Operations.h"
#include "llvm/ADT/STLExtras.h"
#include <tuple>

using namespace llvm;
namespace zkc {
namespace {
using Result = DomainComparison::Result;
std::string fieldOf(StringRef type) {
  auto parts = type.split(':');
  return parts.first == "field" ? parts.second.str() : std::string{};
}

class Analysis {
  PolynomialDomains report;
  using TermKey =
      std::tuple<std::string, std::string, std::string, std::vector<unsigned>>;
  std::map<TermKey, unsigned> interned;
  std::map<std::string, unsigned> exact, lengths;
  std::map<std::tuple<std::string, std::string, unsigned, unsigned>, unsigned>
      classes;

  unsigned term(std::string kind, std::string type, std::string atom = {},
                std::vector<unsigned> args = {}) {
    auto [it, inserted] =
        interned.emplace(TermKey{kind, type, atom, args}, report.terms.size());
    if (inserted)
      report.terms.push_back(
          {std::move(kind), std::move(type), std::move(atom), std::move(args)});
    return it->second;
  }
  unsigned value(const std::string &id) {
    auto found = exact.find(id);
    if (found != exact.end())
      return found->second;
    return exact
        .emplace(id, term("value", report.execution.values.at(id).type, id))
        .first->second;
  }
  unsigned length(const std::string &id) {
    auto found = lengths.find(id);
    if (found != lengths.end())
      return found->second;
    return lengths.emplace(id, term("length", "index", {}, {value(id)}))
        .first->second;
  }
  unsigned half(unsigned n) {
    const auto &t = report.terms[n];
    uint64_t literal;
    if (t.kind == "constant" && !StringRef(t.atom).getAsInteger(10, literal))
      return term("constant", "index", std::to_string(literal / 2));
    return term("half", "index", {}, {n});
  }
  unsigned square(unsigned s) {
    return term("square", report.terms[s].type, {}, {s});
  }
  void transfer(const source::ExecutionOperation &op,
                const protocol::DomainValueContract &facet) {
    const auto &out = op.outputs.at(facet.output);
    const auto &type = report.execution.values.at(out).type;
    using Rule = protocol::DomainValueRule;
    switch (facet.rule) {
    case Rule::Constant:
      exact[out] = term("constant", type, op.attributes.at(0));
      break;
    case Rule::Length:
      exact[out] = length(op.inputs.at(facet.input));
      break;
    case Rule::Multiply: {
      unsigned a = value(op.inputs.at(facet.input));
      unsigned b = value(op.inputs.at(facet.otherInput));
      // Only the needed square law; no broad arithmetic canonicalization.
      if (a == b)
        exact[out] = square(a);
      break;
    }
    case Rule::Divide: {
      unsigned a = value(op.inputs.at(facet.input));
      unsigned b = value(op.inputs.at(facet.otherInput));
      const auto &denominator = report.terms[b];
      if (denominator.kind == "constant" && denominator.atom == "2")
        exact[out] = half(a);
      break;
    }
    }
  }
  unsigned add(PolynomialDomain d) {
    if (!d.field.empty() && !d.convention.empty()) {
      auto [it, inserted] = classes.emplace(
          std::make_tuple(d.field, d.convention, d.size, d.shift),
          classes.size());
      (void)inserted;
      d.congruenceClass = it->second;
    }
    unsigned id = report.domains.size();
    report.domains.push_back(std::move(d));
    return id;
  }

public:
  explicit Analysis(source::Execution execution) {
    report.execution = std::move(execution);
  }
  PolynomialDomains run() {
    for (const auto &path : report.execution.order) {
      const auto &op = report.execution.operations.at(path);
      auto *contracts = protocol::operationContracts(op.callee);
      // A reception is a fresh exact value, never a copy of its sender.
      if (!contracts)
        continue;
      if (contracts->domainValue)
        transfer(op, *contracts->domainValue);
      if (!contracts->coset)
        continue;
      const auto &c = *contracts->coset;
      PolynomialDomain d;
      d.path = path;
      d.operation = op.callee;
      d.role = op.role;
      d.association = "operation";
      d.shiftOrigin = op.inputs.at(c.shiftInput);
      d.sizeOrigin = op.inputs.at(c.sizeInput);
      d.sizeOriginIsLength = c.sizeIsLength;
      d.field = fieldOf(report.execution.values.at(d.shiftOrigin).type);
      if (auto *convention = protocol::cosetConvention(d.field))
        d.convention = convention->identity.str();
      d.shift = value(d.shiftOrigin);
      d.size = c.sizeIsLength ? length(d.sizeOrigin) : value(d.sizeOrigin);
      unsigned domain = add(d);
      if (c.sizeIsLength) {
        DomainUse use;
        use.value = d.sizeOrigin;
        use.domain = domain;
        auto found = report.vectorDomains.find(use.value);
        if (found != report.vectorDomains.end()) {
          use.producerDomain = found->second;
          use.compatibility =
              comparePolynomialDomains(report, found->second, domain);
        } else {
          const auto &v = report.execution.values.at(use.value);
          bool reception =
              v.origin != "$" &&
              !report.execution.operations.at(v.origin).receiver.empty();
          use.compatibility.reasons.push_back(
              reception ? "reception-has-no-trusted-domain"
                        : "no-producer-domain-fact");
        }
        report.uses.push_back(std::move(use));
      }
      if (c.vectorOutput) {
        if (c.foldedOutput) {
          d.association = "fold-result";
          d.size = half(d.size);
          d.shift = square(d.shift);
          domain = add(d);
        }
        const auto &output = op.outputs.at(*c.vectorOutput);
        report.vectorDomains[output] = domain;
        lengths[output] = d.size;
      }
    }
    return std::move(report);
  }
};

DomainComparison compareTerm(const PolynomialDomains &r, unsigned a, unsigned b,
                             StringRef component) {
  if (a == b)
    return {Result::Same, {(component + "-exact-expression").str()}};
  const auto &left = r.terms.at(a), &right = r.terms.at(b);
  if (left.kind == "constant" && right.kind == "constant" &&
      left.type == right.type && left.atom != right.atom)
    return {Result::Different,
            {(component + "-distinct-canonical-constants").str()}};
  if (component == "size" &&
      ((left.kind == "half" && left.arguments.at(0) == b) ||
       (right.kind == "half" && right.arguments.at(0) == a)))
    return {Result::Different, {"size-positive-domain-versus-half"}};
  return {Result::Unknown, {(component + "-equality-unproved").str()}};
}
json::Array strings(const source::Names &names) {
  json::Array result;
  for (const auto &name : names)
    result.push_back(name);
  return result;
}
} // namespace

PolynomialDomains analyzePolynomialDomains(source::Execution e) {
  return Analysis(std::move(e)).run();
}
Expected<PolynomialDomains> inspectPolynomialDomains(const source::Module &m,
                                                     StringRef entry) {
  auto e = source::inspectExecution(m, entry);
  if (!e)
    return e.takeError();
  return analyzePolynomialDomains(std::move(*e));
}
DomainComparison comparePolynomialDomains(const PolynomialDomains &r,
                                          unsigned left, unsigned right) {
  if (left >= r.domains.size() || right >= r.domains.size())
    return {Result::Unknown, {"domain-id-out-of-range"}};
  const auto &a = r.domains[left], &b = r.domains[right];
  if (!a.field.empty() && !b.field.empty() && a.field != b.field)
    return {Result::Different, {"distinct-nominal-fields"}};
  auto size = compareTerm(r, a.size, b.size, "size");
  auto shift = compareTerm(r, a.shift, b.shift, "shift");
  if (size.result == Result::Different)
    return size;
  if (shift.result == Result::Different)
    return shift;
  if (a.field.empty() || b.field.empty())
    return {Result::Unknown, {"nominal-field-unavailable"}};
  if (a.convention.empty() || b.convention.empty())
    return {Result::Unknown, {"root-convention-unavailable"}};
  if (a.convention != b.convention)
    return {Result::Unknown, {"root-convention-compatibility-unproved"}};
  DomainComparison result;
  result.result = size.result == Result::Same && shift.result == Result::Same
                      ? Result::Same
                      : Result::Unknown;
  result.reasons = {"same-nominal-field-and-installed-order"};
  llvm::append_range(result.reasons, size.reasons);
  llvm::append_range(result.reasons, shift.reasons);
  return result;
}

json::Value encodeDomainComparison(const DomainComparison &c) {
  return json::Object{{"relation", c.result == Result::Same        ? "same"
                                   : c.result == Result::Different ? "different"
                                                                   : "unknown"},
                      {"reasons", strings(c.reasons)}};
}
json::Value encodePolynomialDomains(const PolynomialDomains &r) {
  json::Array terms, domains, uses, values;
  for (auto [id, t] : llvm::enumerate(r.terms)) {
    json::Array args;
    for (unsigned a : t.arguments)
      args.push_back(a);
    terms.push_back(json::Object{{"id", id},
                                 {"kind", t.kind},
                                 {"type", t.type},
                                 {"atom", t.atom},
                                 {"arguments", std::move(args)}});
  }
  for (auto [id, d] : llvm::enumerate(r.domains)) {
    json::Object object{{"id", id},
                        {"path", d.path},
                        {"operation", d.operation},
                        {"role", d.role},
                        {"association", d.association},
                        {"field", d.field},
                        {"family", "two-adic-multiplicative-coset"},
                        {"shift", d.shift},
                        {"size", d.size},
                        {"shift_origin", d.shiftOrigin},
                        {"size_origin", d.sizeOrigin},
                        {"size_origin_is_length", d.sizeOriginIsLength}};
    object["congruence_class"] = d.congruenceClass
                                     ? json::Value(*d.congruenceClass)
                                     : json::Value(nullptr);
    if (const auto *c = protocol::cosetConvention(d.field))
      object["convention"] = json::Object{
          {"identity", c->identity},
          {"root_field", c->rootField},
          {"maximal_root", c->maximalRoot},
          {"max_log_size", c->maxLogSize},
          {"root_formula", "embed(maximal_root^(2^max_log_size / size))"},
          {"order", c->order}};
    else
      object["convention"] = nullptr;
    domains.push_back(std::move(object));
  }
  for (const auto &u : r.uses) {
    json::Object object{
        {"value", u.value},
        {"domain", u.domain},
        {"compatibility", encodeDomainComparison(u.compatibility)}};
    object["producer_domain"] = u.producerDomain
                                    ? json::Value(*u.producerDomain)
                                    : json::Value(nullptr);
    uses.push_back(std::move(object));
  }
  for (const auto &[id, v] : r.execution.values) {
    json::Object object{{"id", id},
                        {"name", v.name},
                        {"origin", v.origin},
                        {"role", v.role},
                        {"type", v.type}};
    auto found = r.vectorDomains.find(id);
    object["domain"] = found == r.vectorDomains.end()
                           ? json::Value(nullptr)
                           : json::Value(found->second);
    std::string reason = "entry-has-no-domain-fact";
    json::Array dependencies;
    if (v.origin != "$") {
      const auto &op = r.execution.operations.at(v.origin);
      reason = op.receiver.empty() ? "no-installed-domain-transfer"
                                   : "reception-has-no-trusted-domain";
      if (op.receiver.empty())
        dependencies = strings(op.inputs);
      else
        object["authored_sender_value"] = op.inputs.at(0);
    }
    object["may_depend_on"] = std::move(dependencies);
    if (found == r.vectorDomains.end())
      object["unknown_reason"] = reason;
    values.push_back(std::move(object));
  }
  return json::Object{
      {"format", "zkc.polynomial-domains/1"},
      {"scope",
       "ordered nominal domains, conditional on successful operations"},
      {"terms", std::move(terms)},
      {"domains", std::move(domains)},
      {"uses", std::move(uses)},
      {"values", std::move(values)}};
}
} // namespace zkc
